File: static/js/indicators.js
// Technical Analysis Indicators

// Simple Moving Average (SMA)
function calculateSMA(data, period) {
    const sma = [];
    for (let i = period - 1; i < data.length; i++) {
        const sum = data.slice(i - period + 1, i + 1).reduce((acc, val) => acc + val.close, 0);
        sma.push({
            time: data[i].time,
            value: sum / period
        });
    }
    return sma;
}

// Exponential Moving Average (EMA)
function calculateEMA(data, period) {
    const k = 2 / (period + 1);
    const ema = [];
    let prevEma = data[0].close;

    for (let i = 1; i < data.length; i++) {
        const currentEma = (data[i].close - prevEma) * k + prevEma;
        ema.push({
            time: data[i].time,
            value: currentEma
        });
        prevEma = currentEma;
    }
    return ema;
}

// Relative Strength Index (RSI)
function calculateRSI(data, period = 14) {
    const rsi = [];
    let gains = 0;
    let losses = 0;

    // Calculate initial average gain and loss
    for (let i = 1; i <= period; i++) {
        const change = data[i].close - data[i - 1].close;
        if (change >= 0) {
            gains += change;
        } else {
            losses -= change;
        }
    }

    let avgGain = gains / period;
    let avgLoss = losses / period;

    // Calculate RSI for the rest of the data
    for (let i = period + 1; i < data.length; i++) {
        const change = data[i].close - data[i - 1].close;
        if (change >= 0) {
            avgGain = (avgGain * (period - 1) + change) / period;
            avgLoss = (avgLoss * (period - 1)) / period;
        } else {
            avgGain = (avgGain * (period - 1)) / period;
            avgLoss = (avgLoss * (period - 1) - change) / period;
        }

        const rs = avgGain / avgLoss;
        const rsiValue = 100 - (100 / (1 + rs));

        rsi.push({
            time: data[i].time,
            value: rsiValue
        });
    }

    return rsi;
}

// Moving Average Convergence Divergence (MACD)
function calculateMACD(data, fastPeriod = 12, slowPeriod = 26, signalPeriod = 9) {
    const fastEMA = calculateEMA(data, fastPeriod);
    const slowEMA = calculateEMA(data, slowPeriod);
    const macdLine = [];
    const signalLine = [];
    const histogram = [];

    // Calculate MACD line
    for (let i = 0; i < fastEMA.length; i++) {
        const macdValue = fastEMA[i].value - slowEMA[i].value;
        macdLine.push({
            time: fastEMA[i].time,
            value: macdValue
        });
    }

    // Calculate Signal line (EMA of MACD)
    const signalEMA = calculateEMA(macdLine, signalPeriod);

    // Calculate Histogram
    for (let i = 0; i < signalEMA.length; i++) {
        const histValue = macdLine[i + signalPeriod - 1].value - signalEMA[i].value;
        histogram.push({
            time: signalEMA[i].time,
            value: histValue
        });
    }

    return {
        macd: macdLine.slice(signalPeriod - 1),
        signal: signalEMA,
        histogram: histogram
    };
}

// Bollinger Bands
function calculateBollingerBands(data, period = 20, multiplier = 2) {
    const bands = [];
    const sma = calculateSMA(data, period);

    for (let i = period - 1; i < data.length; i++) {
        const slice = data.slice(i - period + 1, i + 1);
        const standardDeviation = Math.sqrt(
            slice.reduce((acc, val) => acc + Math.pow(val.close - sma[i - period + 1].value, 2), 0) / period
        );

        bands.push({
            time: data[i].time,
            middle: sma[i - period + 1].value,
            upper: sma[i - period + 1].value + (multiplier * standardDeviation),
            lower: sma[i - period + 1].value - (multiplier * standardDeviation)
        });
    }

    return bands;
}

// Volume Weighted Average Price (VWAP)
function calculateVWAP(data) {
    const vwap = [];
    let cumulativeTPV = 0; // Total Price * Volume
    let cumulativeVolume = 0;

    for (let i = 0; i < data.length; i++) {
        const typicalPrice = (data[i].high + data[i].low + data[i].close) / 3;
        const volume = data[i].volume || 0;

        cumulativeTPV += typicalPrice * volume;
        cumulativeVolume += volume;

        vwap.push({
            time: data[i].time,
            value: cumulativeTPV / cumulativeVolume
        });
    }

    return vwap;
}

// Average True Range (ATR)
function calculateATR(data, period = 14) {
    const tr = [];
    const atr = [];

    // Calculate True Range
    for (let i = 1; i < data.length; i++) {
        const high = data[i].high;
        const low = data[i].low;
        const prevClose = data[i - 1].close;

        const tr1 = high - low;
        const tr2 = Math.abs(high - prevClose);
        const tr3 = Math.abs(low - prevClose);

        tr.push(Math.max(tr1, tr2, tr3));
    }

    // Calculate ATR
    let sum = tr.slice(0, period).reduce((acc, val) => acc + val, 0);
    atr.push({
        time: data[period].time,
        value: sum / period
    });

    for (let i = period; i < tr.length; i++) {
        sum = (atr[atr.length - 1].value * (period - 1) + tr[i]) / period;
        atr.push({
            time: data[i + 1].time,
            value: sum
        });
    }

    return atr;
}

// Stochastic Oscillator
function calculateStochastic(data, kPeriod = 14, dPeriod = 3) {
    const stoch = [];
    const kLine = [];
    const dLine = [];

    for (let i = kPeriod - 1; i < data.length; i++) {
        const slice = data.slice(i - kPeriod + 1, i + 1);
        const highestHigh = Math.max(...slice.map(candle => candle.high));
        const lowestLow = Math.min(...slice.map(candle => candle.low));
        const currentClose = data[i].close;

        const k = ((currentClose - lowestLow) / (highestHigh - lowestLow)) * 100;
        kLine.push({
            time: data[i].time,
            value: k
        });
    }

    // Calculate D line (SMA of K line)
    const d = calculateSMA(kLine, dPeriod);
    dLine.push(...d);

    return {
        k: kLine,
        d: dLine
    };
}

// Export all indicator functions
export {
    calculateSMA,
    calculateEMA,
    calculateRSI,
    calculateMACD,
    calculateBollingerBands,
    calculateVWAP,
    calculateATR,
    calculateStochastic
};

File: src/indicators.py
import numpy as np
import pandas as pd
from typing import Dict, List, Optional, Tuple, Union

class TechnicalIndicators:
    """Technical analysis indicators."""
    
    @staticmethod
    def sma(data: pd.Series, period: int) -> pd.Series:
        """Calculate Simple Moving Average.
        
        Args:
            data: Price series
            period: SMA period
            
        Returns:
            Series with SMA values
        """
        return data.rolling(window=period).mean()
        
    @staticmethod
    def ema(data: pd.Series, period: int) -> pd.Series:
        """Calculate Exponential Moving Average.
        
        Args:
            data: Price series
            period: EMA period
            
        Returns:
            Series with EMA values
        """
        return data.ewm(span=period, adjust=False).mean()
        
    @staticmethod
    def rsi(data: pd.Series, period: int = 14) -> pd.Series:
        """Calculate Relative Strength Index.
        
        Args:
            data: Price series
            period: RSI period
            
        Returns:
            Series with RSI values
        """
        delta = data.diff()
        gain = (delta.where(delta > 0, 0)).rolling(window=period).mean()
        loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()
        rs = gain / loss
        return 100 - (100 / (1 + rs))
        
    @staticmethod
    def macd(data: pd.Series, fast: int = 12, slow: int = 26,
             signal: int = 9) -> Tuple[pd.Series, pd.Series, pd.Series]:
        """Calculate MACD.
        
        Args:
            data: Price series
            fast: Fast period
            slow: Slow period
            signal: Signal period
            
        Returns:
            Tuple of (MACD line, Signal line, Histogram)
        """
        exp1 = data.ewm(span=fast, adjust=False).mean()
        exp2 = data.ewm(span=slow, adjust=False).mean()
        macd = exp1 - exp2
        signal_line = macd.ewm(span=signal, adjust=False).mean()
        hist = macd - signal_line
        return macd, signal_line, hist
        
    @staticmethod
    def bollinger_bands(data: pd.Series, period: int = 20,
                       std: float = 2) -> Tuple[pd.Series, pd.Series, pd.Series]:
        """Calculate Bollinger Bands.
        
        Args:
            data: Price series
            period: BB period
            std: Standard deviation multiplier
            
        Returns:
            Tuple of (Upper band, Middle band, Lower band)
        """
        middle = data.rolling(window=period).mean()
        std_dev = data.rolling(window=period).std()
        upper = middle + (std_dev * std)
        lower = middle - (std_dev * std)
        return upper, middle, lower
        
    @staticmethod
    def stochastic(data: pd.DataFrame, k_period: int = 14,
                  d_period: int = 3) -> Tuple[pd.Series, pd.Series]:
        """Calculate Stochastic Oscillator.
        
        Args:
            data: DataFrame with high, low, close prices
            k_period: %K period
            d_period: %D period
            
        Returns:
            Tuple of (%K line, %D line)
        """
        low_min = data['low'].rolling(window=k_period).min()
        high_max = data['high'].rolling(window=k_period).max()
        k = 100 * ((data['close'] - low_min) / (high_max - low_min))
        d = k.rolling(window=d_period).mean()
        return k, d
        
    @staticmethod
    def atr(data: pd.DataFrame, period: int = 14) -> pd.Series:
        """Calculate Average True Range.
        
        Args:
            data: DataFrame with high, low, close prices
            period: ATR period
            
        Returns:
            Series with ATR values
        """
        high = data['high']
        low = data['low']
        close = data['close']
        
        tr1 = high - low
        tr2 = abs(high - close.shift())
        tr3 = abs(low - close.shift())
        
        tr = pd.concat([tr1, tr2, tr3], axis=1).max(axis=1)
        return tr.rolling(window=period).mean()
        
    @staticmethod
    def adx(data: pd.DataFrame, period: int = 14) -> Tuple[pd.Series, pd.Series, pd.Series]:
        """Calculate Average Directional Index.
        
        Args:
            data: DataFrame with high, low, close prices
            period: ADX period
            
        Returns:
            Tuple of (ADX, +DI, -DI)
        """
        high = data['high']
        low = data['low']
        close = data['close']
        
        # Calculate True Range
        tr1 = high - low
        tr2 = abs(high - close.shift())
        tr3 = abs(low - close.shift())
        tr = pd.concat([tr1, tr2, tr3], axis=1).max(axis=1)
        
        # Calculate Directional Movement
        up_move = high - high.shift()
        down_move = low.shift() - low
        
        plus_dm = np.where((up_move > down_move) & (up_move > 0), up_move, 0)
        minus_dm = np.where((down_move > up_move) & (down_move > 0), down_move, 0)
        
        # Calculate Smoothed Averages
        tr_smoothed = tr.rolling(window=period).sum()
        plus_di = 100 * pd.Series(plus_dm).rolling(window=period).sum() / tr_smoothed
        minus_di = 100 * pd.Series(minus_dm).rolling(window=period).sum() / tr_smoothed
        
        # Calculate ADX
        dx = 100 * abs(plus_di - minus_di) / (plus_di + minus_di)
        adx = dx.rolling(window=period).mean()
        
        return adx, plus_di, minus_di
        
    @staticmethod
    def ichimoku(data: pd.DataFrame, tenkan: int = 9, kijun: int = 26,
                senkou_span_b: int = 52, displacement: int = 26) -> Dict[str, pd.Series]:
        """Calculate Ichimoku Cloud.
        
        Args:
            data: DataFrame with high, low, close prices
            tenkan: Tenkan-sen period
            kijun: Kijun-sen period
            senkou_span_b: Senkou Span B period
            displacement: Displacement period
            
        Returns:
            Dictionary with Ichimoku components
        """
        high = data['high']
        low = data['low']
        
        # Calculate Tenkan-sen (Conversion Line)
        tenkan_high = high.rolling(window=tenkan).max()
        tenkan_low = low.rolling(window=tenkan).min()
        tenkan_sen = (tenkan_high + tenkan_low) / 2
        
        # Calculate Kijun-sen (Base Line)
        kijun_high = high.rolling(window=kijun).max()
        kijun_low = low.rolling(window=kijun).min()
        kijun_sen = (kijun_high + kijun_low) / 2
        
        # Calculate Senkou Span A (Leading Span A)
        senkou_span_a = ((tenkan_sen + kijun_sen) / 2).shift(displacement)
        
        # Calculate Senkou Span B (Leading Span B)
        senkou_high = high.rolling(window=senkou_span_b).max()
        senkou_low = low.rolling(window=senkou_span_b).min()
        senkou_span_b = ((senkou_high + senkou_low) / 2).shift(displacement)
        
        # Calculate Chikou Span (Lagging Span)
        chikou_span = data['close'].shift(-displacement)
        
        return {
            'tenkan_sen': tenkan_sen,
            'kijun_sen': kijun_sen,
            'senkou_span_a': senkou_span_a,
            'senkou_span_b': senkou_span_b,
            'chikou_span': chikou_span
        }
        
    @staticmethod
    def fibonacci_retracement(high: float, low: float) -> Dict[str, float]:
        """Calculate Fibonacci Retracement levels.
        
        Args:
            high: Highest price
            low: Lowest price
            
        Returns:
            Dictionary with retracement levels
        """
        diff = high - low
        return {
            '0.0': low,
            '0.236': low + 0.236 * diff,
            '0.382': low + 0.382 * diff,
            '0.5': low + 0.5 * diff,
            '0.618': low + 0.618 * diff,
            '0.786': low + 0.786 * diff,
            '1.0': high
        }
        
    @staticmethod
    def pivot_points(data: pd.DataFrame) -> Dict[str, float]:
        """Calculate Pivot Points.
        
        Args:
            data: DataFrame with high, low, close prices
            
        Returns:
            Dictionary with pivot points
        """
        high = data['high'].iloc[-1]
        low = data['low'].iloc[-1]
        close = data['close'].iloc[-1]
        
        pp = (high + low + close) / 3
        
        return {
            'pp': pp,
            'r1': 2 * pp - low,
            'r2': pp + (high - low),
            'r3': high + 2 * (pp - low),
            's1': 2 * pp - high,
            's2': pp - (high - low),
            's3': low - 2 * (high - pp)
        }

def moving_average(data, window):
    return sum(data[-window:]) / window if len(data) >= window else 0

def rsi(data, period=14):
    return 50

class Indicator:
    pass

def calculateSMA(data, period):
    return [0] * len(data)

def calculateEMA(data, period):
    return [0] * len(data)

def calculateRSI(data, period):
    return [0] * len(data)

def calculateMACD(data, fast_period, slow_period, signal_period):
    return [0] * len(data), [0] * len(data), [0] * len(data)

def calculateBollingerBands(data, period, std_dev):
    return [0] * len(data), [0] * len(data), [0] * len(data)

def calculateVWAP(data, volume):
    return [0] * len(data)

def calculateATR(data, period):
    return [0] * len(data)

def calculateStochastic(data, period):
    return [0] * len(data) 